### PR TITLE
CLB: Cadira, Caller of the Small + 4 cards

### LINKED ISSUE #588: Archivist of Oghma

### LINKED ISSUE #589: Ascend from Avernus

### LINKED ISSUE #590: Astarion's Thirst

### LINKED ISSUE #591: Baldur's Gate

### LINKED ISSUE #592: Cadira, Caller of the Small

### PR DESCRIPTION
**Pull request by Suthro**
CLB: Contains scripts for five CBL cards —

- closes #588 
- closes #589 
- closes #590 
- closes #591 
- closes #592 

Contains 1 token script:
- 1/1 white Rabbit creature token

### CODE REVIEW

**Review by tool4ever — DISMISSED** [review]

File: forge-gui/res/cardsfolder/upcoming/archivist_of_oghma.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Name:Archivist of Oghma
+ManaCost:1 W
+Types:Creature Halfling Cleric
+PT:2/2
+K:Flash
+T:Mode$ SearchedLibrary | ValidPlayer$ Player.Opponent | SearchOwnLibrary$ True | Execute$ TrigGainLife | TriggerZones$ Battlefield | TriggerDescription$ Whenever an opponent searches their library, you gain 1 life and draw a card.
+SVar:TrigGainLife:DB$ GainLife | LifeAmount$ 1 | SubAbility$ DBDraw
+SVar:DBDraw:DB$ Draw
+Oracle:Flash\nWhenever an opponent searches their library, you gain 1 life and draw a card.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/archivist_of_oghma.txt:9]: `DeckHas:Ability$LifeGain`

File: forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Name:Ascend from Avernus
+ManaCost:X W W W
+Types:Sorcery
+A:SP$ ChangeZoneAll | Origin$ Graveyard | Destination$ Battlefield | ChangeType$ Creature.YouOwn+cmcLEX,Planeswalker.YouOwn+cmcLEX | SubAbility$ DBChange | SpellDescription$ Return all creature and planeswalker cards with mana value X or less from your graveyard to the battlefield. Exile CARDNAME.
+SVar:DBChange:DB$ ChangeZone | Origin$ Stack | Destination$ Exile
+SVar:X:Count$xPaid
+DeckHas:Ability$Graveyard
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt:7]: how about:
`DeckHints:Type$Planeswalker & Ability$Discard|Sacrifice`

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Name:Astarion's Thirst
+ManaCost:3 B
+Types:Instant
+A:SP$ ChangeZone | ValidTgts$ Creature | TgtPrompt$ Select target creature | Origin$ Battlefield | Destination$ Exile | RememberChanged$ True | SubAbility$ DBPutCounter | SpellDescription$ Exile target creature. Put X +1/+1 counters on a commander creature you control, where X is the power of the creature exiled this way.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:4]: FYI `TgtPrompt$ Select target creature | ` is not needed - Forge will parse this without help

Fine if you want to leave it though

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+SVar:DBPutCounter:DB$ PutCounter | Choices$ Creature.YouCtrl+IsCommander | CounterType$ P1P1 | CounterNum$ X | SubAbility$ DBCleanup
+SVar:DBCleanup:DB$ Cleanup | ClearRemembered$ True
+SVar:X:RememberedLKI$CardPower
+Oracle:Exile target creature. Put X +1/+1 counters on a commander creature you control, where X is the power of the creature exiled this way.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:8]: DeckHas:Ability$Counters

File: forge-gui/res/cardsfolder/upcoming/archivist_of_oghma.txt (hunk 1 above)
Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/archivist_of_oghma.txt]: Added life gain deck flag

File: forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt (hunk 2 above)
Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt:7]: Added planeswalker type deck hint
Added discard/sacrifice ability deck hint

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt (hunk 4 above)
Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: Added counters ability deck flag

File: forge-gui/res/cardsfolder/upcoming/cadira_caller_of_the_small.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Name:Cadira, Caller of the Small
+ManaCost:1 G W
+Types:Legendary Creature Orc Ranger
+PT:3/3
+K:Trample
+T:Mode$ DamageDone | ValidSource$ Card.Self | ValidTarget$ Player | CombatDamage$ True | Execute$ TrigToken | TriggerDescription$ Whenever CARDNAME deals combat damage to a player, for each token you control, create a 1/1 white Rabbit creature token.
+SVar:TrigToken:DB$ Token | TokenScript$ w_1_1_rabbit | TokenOwner$ You | TokenAmount$ X
+SVar:X:Count$Valid Permanent.token+YouCtrl
+DeckHas:Ability$Token
```
[the remark this answers is not in the document]
Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/cadira_caller_of_the_small.txt]: Added rabbit type deck flag
Added token ability `DeckNeeds` flag

File: forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+Name:Ascend from Avernus
+ManaCost:X W W W
+Types:Sorcery
+A:SP$ ChangeZoneAll | Origin$ Graveyard | Destination$ Battlefield | ChangeType$ Creature.YouOwn+cmcLEX,Planeswalker.YouOwn+cmcLEX | SubAbility$ DBChange | SpellDescription$ Return all creature and planeswalker cards with mana value X or less from your graveyard to the battlefield. Exile CARDNAME.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt:4]: I think you should trim `Exile CARDNAME.` from this SpellDescription$ and add it to a SpellDescription in line 5

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Name:Astarion's Thirst
+ManaCost:3 B
+Types:Instant
+A:SP$ ChangeZone | ValidTgts$ Creature | TgtPrompt$ Select target creature | Origin$ Battlefield | Destination$ Exile | RememberChanged$ True | SubAbility$ DBPutCounter | SpellDescription$ Exile target creature. Put X +1/+1 counters on a commander creature you control, where X is the power of the creature exiled this way.
+SVar:DBPutCounter:DB$ PutCounter | Choices$ Creature.YouCtrl+IsCommander | CounterType$ P1P1 | CounterNum$ X | SubAbility$ DBCleanup
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:5]: How does the StackDescription look for this on cast?

Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:5]: The current script produces this stack description:

"Exile {targeted}. they puts zero +1/+1 counters on a Creature.YouCtrl+IsCommander."

I took some time testing the StackDescriptions further, changing the end of line 4 to the following:

`StackDescription$ Exile {c:Targeted}. | SpellDescription$ StackDescription`

And adding a StackDescription to line 5 as follows:

`StackDescription$ Put X +1/+1 counters on a commander creature you control, where X is the power of the creature exiled this way.`

With these changes, I arrived at a stack description that reads:

"Exile {targeted}. Put X +1/+1 counters on a commander creature you control."

Is this the end goal we're seeking? Or are we looking to have the stack description say "<player> puts X +1/+1 counters on a commander creature <they> control"?

File: forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt (hunk 6 above)
Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/ascend_from_avernus.txt]: Moved `Exile CARDNAME` clause from line 4 to line 5

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt (hunk 7 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:5]: Line 4 should work fine with just `SpellDescription$ Exile CARDNAME.` Forge should be able to parse this without any trouble.
Line 5 is the one Forge would struggle parsing, and I think your solution is fine.

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: Line 5 you probably want:
```
StackDescription$ SpellDescription | SpellDescription$ Put X +1/+1 counters on a commander creature you control, where X is the power of the creature exiled this way.
```

As a general rule you want to check the card detail pane and the stack. The card detail pane should match the oracle, though re-ordering of triggers, statics, etc. is fine. The stack should at least make sense. If you see lots of unparsed output then you may want to reach for the StackDescription or ping me about it.

Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: I see what went wrong then, I must have swapped the order of StackDesc$ and SpellDesc$. Appreciate the extra set of eyes

File: forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt [hunk 8]
```diff
@@ -2,6 +2,6 @@ Name:Astarion's Thirst
 ManaCost:3 B
 Types:Instant
 A:SP$ ChangeZone | ValidTgts$ Creature | TgtPrompt$ Select target creature | Origin$ Battlefield | Destination$ Exile | SubAbility$ DBPutCounter | StackDescription$ Exile {c:Targeted}. | SpellDescription$ StackDescription
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt:4]: I don't believe this will look right in the card detail pane.

Comment (Suthro) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: Looking at the card detail pane and the stack description, I think I've come to understand that the `{c:Targeted}` tag is causing some problem when used in the `SpellDescription$` string. I've revised the spell description to state the full spell text as written in oracle, but kept in the `StackDescription$` string in line 5. Admittedly there is probably a more efficient way to do this, but it seems that this combination is now successfully presenting the correct stack description and card detail text in testing.

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: The tokens (like `{c: whatever}`) are only for StackDescription... they aren't parsed for SpellDescription

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/astarions_thirst.txt]: In terms of efficiency it'd have been easier to leave line 5 alone and just set line 4 `SpellDescription$` to the first sentence of the oracle text

But this works!

**Review by Northmoc — APPROVED** [review]
👍

**Pull request merged**